roles/setup_efm/tasks/setup_efm.yml

```diff
@@ -13,20 +13,91 @@
   become: yes
   when:
     - use_hostname
-  delegate_to: "{{ item.value.public_ip }}"
 
 - name: Remove before we start as per force_install
   block:
     - include_tasks: rm_efm_install_config.yml
   become: yes
   when: force_efm_install
-  delegate_to: "{{ item.value.public_ip }}"
+- name: Prepare etc_hosts_lists
+  set_fact:
+    etc_hosts_lists: "{{ etc_hosts_lists + [node.private_ip + ' ' + node.hostname] }}"
+  loop: "{{ lookup('pg_sr_cluster_nodes', wantlist=True) }}"
+  loop_control:
+    loop_var: node
+  when: use_hostname
+  no_log: "{{ disable_logging }}"
+
+- name: Get the primary information
+  set_fact:
+    primary_public_ip: "{{ node.ansible_host }}"
+    primary_private_ip: "{{ node.private_ip }}"
+  loop: "{{ lookup('pg_sr_cluster_nodes', wantlist=True) }}"
+  loop_control:
+    loop_var: node
+  when:  node.node_type == 'primary'
+  no_log: "{{ disable_logging }}"
+
+- name: Set efm install version
+  set_fact:
+    efm_install_version : "{{ efm_version | replace('.','') }}"
+  
+- name: Prepare efm node list
+  set_fact:
+    efm_nodes_list: "{{ efm_nodes_list + node.private_ip + ':7800 ' }}"
+  when:
+    - node.node_type in supported_node_type
+    - not use_hostname
+  loop: "{{ lookup('pg_sr_cluster_nodes', wantlist=True) }}"
+  loop_control:
+    loop_var: node
+  no_log: "{{ disable_logging }}"  
+
+- name: Prepare efm node list based on hostname
+  set_fact:
+    efm_nodes_list: "{{ efm_nodes_list + node.hostname + ':7800 ' }}"
+  when:
+    - node.node_type in supported_node_type
+    - node.hostname is defined
+    - use_hostname
+  loop: "{{ lookup('pg_sr_cluster_nodes', wantlist=True) }}"
+  loop_control:
+    loop_var: node
+  no_log: "{{ disable_logging }}"
+
+- name: Prepare host ip lists based on allow_node_ips
+  include_tasks: prepare_hba_value_list.yml
+  loop: "{{ lookup('pg_sr_cluster_nodes', wantlist=True) }}"
+  loop_control:
+    loop_var: node
+  no_log: "{{ disable_logging }}"
+
+- name: Ensure we have efm user in primary database
+  import_tasks: create_efm_user.yml
+  run_once: true
+  no_log: "{{ disable_logging }}"
+  delegate_to: "{{ primary_public_ip }}"
```

Comment:
Could we just check if the current node is a primary node instead of executing this for each nodes ?

Reply:
Sure